docs/internals/annotations_resolving.md

```diff
@@ -0,0 +1,267 @@
+!!! note
+    This section is part of the *internals* documentation, and is partly targeted to contributors.
+
+Pydantic heavily relies on [type hints][type hint] at runtime to build schemas for validation, serialization, etc.
+
+While type hints were primarily introduced for static type checkers (such as [Mypy] or [Pyright]), they are
+accessible (and sometimes evaluated) at runtime. This means that the following would fail at runtime,
+because `Node` has yet to be defined in the current module:
+
+```python test="skip" lint="skip"
+class Node:
+    """Binary tree node."""
+
+    # NameError: name 'Node' is not defined:
+    def __init__(self, l: Node, r: Node) -> None:
+        self.left = l
+        self.right = r
+```
+
+To circumvent this issue, forward references can be used (by wrapping the annotation into quotes).
+
+In Python 3.7, [PEP 563] introduced the concept of _postponed evaluation of annotations_, meaning
+with the `from __future__ import annotations` [future statement], type hints are stringified by default:
+
+```python requires="3.12" lint="skip"
+from __future__ import annotations
+
+from pydantic import BaseModel
+
+
+class Foo(BaseModel):
+    f: MyType
+    # Without the future import, equivalent to:
+    # f: 'MyType'
+
+
+type MyType = int
+
+print(Foo.__annotations__)
+#> {'f': 'MyType'}
+```
+
+## The challenges of runtime evaluation
+
+Static type checkers make use of the <abbr title="Abstract Syntax Tree">AST</abbr> to analyze the defined annotations.
+Regarding the previous example, this has the benefit of being able to understand what `MyType` refers to when analyzing
+the class definition of `Foo`, even if it isn't yet defined at runtime.
+
+However, for runtime tools such as Pydantic, it is more challenging to correctly resolve these forward annotations.
+The Python standard library provides some tools to do so ([`typing.get_type_hints()`][typing.get_type_hints],
+[`inspect.get_annotations()`][inspect.get_annotations]), but they come with some limitations. Thus, they are
+being re-implemented in Pydantic with improved support for edge cases.
+
+For a long time, Pydantic has tried to support many cases, resulting in a messy annotations evaluation logic
+(which would allow invalid use cases from a static type checking perspective to work). In 2.10, the internal logic
```

Comment:
So, I fully agree with what you're saying here.

That being said, I don't think we should be quite this negative in the docs. Maybe something like:

As Pydantic as grown, it's adapted to support many edge cases requiring irregular patterns for annotation evaluation. Some of these use cases aren't necessarily sound from a static type checking perspective. In v2.10, the internal namespace management logic was refactored in an attempt to simplify and standardize annotation evaluation. Admittedly, backwards compatibility posed some challenges, and there is still some noticeable scar tissue in the codebase because of this. There's hope that...